### PR TITLE
Feat: add support for Heredocs in Databricks Python UDFs

### PR DESCRIPTION
**Pull request by viethungle-vt1401**
This PR adds support for Heredocs in the `AS dollar_quoted_string` clause for Python user-defined functions in [Databricks](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-create-sql-function.html#create-python-functions).

### CODE REVIEW

File: tests/dialects/test_databricks.py [hunk 1]
```diff
@@ -28,6 +28,13 @@ def test_databricks(self):
             "SELECT * FROM sales UNPIVOT EXCLUDE NULLS (sales FOR quarter IN (q1 AS `Jan-Mar`))"
         )
 
+        self.validate_all(
+            "CREATE FUNCTION greet(name STRING) RETURNS STRING LANGUAGE PYTHON AS $$def greet(name):\n\t return f'Hello {name}!'$$",
+            write={
+                "snowflake": "CREATE FUNCTION greet(name TEXT) RETURNS TEXT LANGUAGE PYTHON AS 'def greet(name):\n\t return f\\'Hello {name}!\\''"
+            },
+        )
+
```
Comment (georgesittas) [inline tests/dialects/test_databricks.py:37]: This test is no good, we're not testing the roundtrip from Databricks to Databricks. Have you double-checked that this is valid Snowflake?

File: sqlglot/parser.py [hunk 2]
```diff
@@ -5475,6 +5476,37 @@ def _parse_comprehension(
...
+        # Note: we've already matched a $ token before reaching here
+        self._match(TokenType.PARAMETER)
+        tag_start = self._prev
+
+        while not self._match(TokenType.PARAMETER):
+            self._advance()
+            if not self._curr:
+                self.raise_error("No closing $ found.")
+
+        tag_text = self._find_sql(tag_start, self._prev)
```
Comment (georgesittas) [inline sqlglot/parser.py:5489]: Why is there a while loop here? I think this section can be simplified, don't think we need `_find_sql` here, you can just get the token text directly from the tag's token, if there's a tag.

File: sqlglot/parser.py [hunk 3]
```diff
@@ -1362,9 +1362,10 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
             # exp.Properties.Location.POST_SCHEMA ("schema" here is the UDF's type signature)
             extend_props(self._parse_properties())
 
-            self._match(TokenType.ALIAS)
-
-            if self._match(TokenType.COMMAND):
+            if self._match(TokenType.ALIAS) and self._match(TokenType.PARAMETER, advance=False):
```
[the remark this answers is not in the document]
Comment (georgesittas) [reply sqlglot/parser.py:1365]: Also, it's hard to understand why we're matching a `PARAMETER` token when reading this section. Is this syntax only valid for Databricks? Maybe we should be matching a dollar text instead using `self._match_text_seq("$")` or something?

File: sqlglot/parser.py [hunk 4]
```diff
@@ -5475,6 +5476,37 @@ def _parse_comprehension(
...
+
+        tag_start = None
+        while True:
+            if self._match(TokenType.PARAMETER, advance=False):
+                if tag_start and self._find_sql(tag_start, self._curr) == tag_text:
+                    this = self._find_sql(heredoc_start, heredoc_end)
+                    self._advance()
+
+                    return self.expression(exp.RawString, this=this)
+                else:
```
Comment (georgesittas) [inline sqlglot/parser.py:5501]: This `else` here is unnecessary

File: sqlglot/parser.py [hunk 5]
```diff
@@ -5475,6 +5476,37 @@ def _parse_comprehension(
...
+        tag_text = self._find_sql(tag_start, self._prev)
+        heredoc_start = self._curr
+        heredoc_end = self._prev
+
+        tag_start = None
+        while True:
+            if self._match(TokenType.PARAMETER, advance=False):
+                if tag_start and self._find_sql(tag_start, self._curr) == tag_text:
+                    this = self._find_sql(heredoc_start, heredoc_end)
+                    self._advance()
```
Comment (georgesittas) [inline sqlglot/parser.py:5498]: Can this be simplified? I'd expect this section to just contain some lookahead checks. For example: "if there's a $ followed by a token that matches the tag text (if any), _and_ there's another dollar, we're at the heredoc's end and only then we should capture the raw text with `_find_sql`".

**Review by georgesittas — COMMENTED** [review]
Left some comments to consider. I did a quick first pass so let me know if I missed something in your logic.

File: sqlglot/parser.py [hunk 6]
```diff
@@ -1362,9 +1362,12 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
             # exp.Properties.Location.POST_SCHEMA ("schema" here is the UDF's type signature)
             extend_props(self._parse_properties())
 
-            self._match(TokenType.ALIAS)
+            alias_ = self._match(TokenType.ALIAS)
```
Comment (tobymao) [inline sqlglot/parser.py:1365]: you can just put this inline in the if statement since you don't use alias_ anywhere

File: sqlglot/parser.py [hunk 7]
```diff
@@ -5475,6 +5478,31 @@ def _parse_comprehension(
             condition=condition,
         )
 
+    def _parse_heredoc(self) -> exp.RawString:
+        tag_texts = ["$", "$"]
```
Comment (tobymao) [inline sqlglot/parser.py:5482]: what about something like

tags = []

if self._match_text_seq("$"):
  tags.append("$")
else: 
   return None

tags.append(self._advance_any().text)

if tags[-1] != "$":
   if self._match_text("$"):
      tags.append("$")
   else:
      self.raise_error(...)

File: sqlglot/parser.py [hunk 8]
```diff
@@ -5475,6 +5478,31 @@ def _parse_comprehension(
...
+
+        if not self._match_text_seq("$"):
+            tag_texts.insert(1, self._curr.text)
+            self._advance()
+
+        self._match_text_seq("$")
+
+        heredoc_start = self._curr
+
+        while True:
```
Comment (tobymao) [inline sqlglot/parser.py:5494]: while self._curr

File: sqlglot/parser.py [hunk 9]
```diff
@@ -5550,10 +5578,17 @@ def _match_texts(self, texts, advance=True):
             return True
         return False
 
-    def _match_text_seq(self, *texts, advance=True):
+    def _match_text_seq(self, *texts, advance=True, upper_case=True):
```
Comment (tobymao) [inline sqlglot/parser.py:5581]: i don't think we need this upper_case flag, it's a bit edge casey, should be fine to just not have this check

File: tests/dialects/test_databricks.py (hunk 1 above)
Comment (tobymao) [reply tests/dialects/test_databricks.py:37]: you should also test the here docs version

File: sqlglot/parser.py (hunk 6 above)
Comment (georgesittas) [reply sqlglot/parser.py]: I actually suggested we don't put that in the if haha.. the reasoning was to make this a bit more clear, i.e. that we'll always consume the alias regardless of the taken branch

Comment (georgesittas) [reply sqlglot/parser.py]: But it's fine

File: sqlglot/generator.py [hunk 10]
```diff
@@ -77,6 +77,7 @@ class Generator:
         exp.ExecuteAsProperty: lambda self, e: self.naked_property(e),
         exp.ExternalProperty: lambda self, e: "EXTERNAL",
         exp.HeapProperty: lambda self, e: "HEAP",
+        exp.Heredoc: lambda self, e: f"${e.args['tag'] or ''}$\n{e.this}\n${e.args['tag'] or ''}$",
```
Comment (tobymao) [inline sqlglot/generator.py:80]: make this e.args("tag") and move this to a first class function

def heredoc_sql ...

File: sqlglot/parser.py [hunk 11]
```diff
@@ -5475,6 +5476,53 @@ def _parse_comprehension(
...
+                tag_text = self._curr.text
+                tags.append(tag_text)
+                self._advance()
+            else:
+                tag_text = None
+                tags.append("$")
+        else:
+            self.raise_error("No closing $ found")
+
+        if tag_text:
```
Comment (tobymao) [inline sqlglot/parser.py:5498]: tag_text can be undefined here because the raise_error on line 5496 doesn't necessarily break

File: sqlglot/parser.py [hunk 12]
```diff
@@ -5475,6 +5476,53 @@ def _parse_comprehension(
...
+            if self._is_connected() and self._match_text_seq("$"):
+                tags.append("$")
+            else:
+                self.raise_error("No closing $ found")
+
+        heredoc_start = self._curr
+
+        while self._curr:
+            index = self._index
```
Comment (tobymao) [inline sqlglot/parser.py:5507]: shouldn't this index be outside of the for loop? perhaps it should be at line 5480?

File: sqlglot/parser.py [hunk 13]
```diff
@@ -5475,6 +5476,53 @@ def _parse_comprehension(
...
+            heredoc_end = self._prev
+            tags_matched = True
+
+            for tag in tags:
```
Comment (tobymao) [inline sqlglot/parser.py:5511]: why don't you do match_text_seq here?

Comment (viethungle-vt1401) [reply sqlglot/parser.py:5511]: `match_text_seq` would convert everything to upper case, meanwhile we want the beginning and closing tags to be the same, case sensitive (e.g. `$foo$` and `$FOO$` shouldn't match)

File: sqlglot/parser.py (hunk 12 above)
Comment (viethungle-vt1401) [reply sqlglot/parser.py:5507]: since I'm not using `_match_text_seq`, the `index` keeps track of `self._curr` is before we try to match subsequent tokens. If the matching fails at any point, we return to `index` before calling `self._advance()` to advance `self._curr` (`index` here serves a similar purpose to `index` in `_match_text_seq`)

File: sqlglot/parser.py (hunk 13 above)
Comment (tobymao) [reply sqlglot/parser.py:5519]: as i said, i don't think this edge case matters much, so let's just use match_text_seq

File: sqlglot/generator.py [hunk 14]
```diff
@@ -77,6 +77,7 @@ class Generator:
         exp.ExecuteAsProperty: lambda self, e: self.naked_property(e),
         exp.ExternalProperty: lambda self, e: "EXTERNAL",
         exp.HeapProperty: lambda self, e: "HEAP",
+        exp.Heredoc: lambda self, e: self.heredoc_sql(e),
```
Comment (tobymao) [inline sqlglot/generator.py:80]: delete this, it's unnecessary

File: sqlglot/parser.py [hunk 15]
```diff
@@ -5475,6 +5484,44 @@ def _parse_comprehension(
...
+        else:
+            return None
+
+        if self._curr and self._is_connected():
+            if not self._match_text_seq("$"):
+                tag_text = self._curr.text
+                tags.append(tag_text.upper())
+                self._advance()
+            else:
+                tags.append("$")
```
Comment (georgesittas) [inline sqlglot/parser.py:5502]: Perhaps you can replace this whole if-else block with `tags.append(self._advance() and self._prev.text.upper())`?

File: sqlglot/parser.py [hunk 16]
```diff
@@ -5475,6 +5484,44 @@ def _parse_comprehension(
...
+    def _parse_heredoc(self) -> t.Optional[exp.Heredoc]:
+        tags = []
+        tag_text = None
+
+        if self._match_text_seq("$"):
+            tags.append("$")
+        else:
+            return None
+
+        if self._curr and self._is_connected():
```
Comment (georgesittas) [inline sqlglot/parser.py:5496]: I think `self._curr` isn't needed here? Just `_is_connected()` seems to be enough.

File: sqlglot/parser.py [hunk 17]
```diff
@@ -5475,6 +5484,44 @@ def _parse_comprehension(
...
+            if not self._match_text_seq("$"):
+                tag_text = self._curr.text
+                tags.append(tag_text.upper())
+                self._advance()
+            else:
+                tags.append("$")
+        else:
+            self.raise_error("No closing $ found")
+
+        if tag_text:
```
Comment (georgesittas) [inline sqlglot/parser.py:5506]: And so instead of checking `tag_text` here, the proxy would be if the last tag appended is _not_ a single dollar?

File: sqlglot/parser.py [hunk 18]
```diff
@@ -5475,6 +5484,44 @@ def _parse_comprehension(
...
+            if self._is_connected() and self._match_text_seq("$"):
+                tags.append("$")
```
Comment (georgesittas) [inline sqlglot/parser.py:5508]: Do we need the `_is_connected` check here?

File: sqlglot/parser.py (hunk 15 above)
Comment (georgesittas) [reply sqlglot/parser.py:5502]: @tobymao wdyt?

File: sqlglot/parser.py [hunk 19]
```diff
@@ -1362,9 +1362,18 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
...
-
-            if self._match(TokenType.COMMAND):
+            heredoc_matched = False
+            if self._match(TokenType.ALIAS):
+                if self._match_text_seq("$", advance=False):
+                    expression = self._parse_heredoc()
+                    heredoc_matched = True
+                elif self._match(TokenType.HEREDOC_STRING):
+                    expression = self.expression(exp.Heredoc, this=self._prev.text)
+                    heredoc_matched = True
```
Comment (georgesittas) [inline sqlglot/parser.py:1372]: should we consolidate these two branches by handling it all in `_parse_heredoc` and setting `heredoc_matched` to true if we get back a non-`None` result? just trying to see if we can simplify it a bit

File: sqlglot/parser.py (hunk 15 above)
Comment (tobymao) [reply sqlglot/parser.py:5502]: sure

File: sqlglot/parser.py (hunk 18 above)
Comment (viethungle-vt1401) [reply sqlglot/parser.py]: This is to catch invalid tags like $foo $

File: sqlglot/parser.py (hunk 15 above)
Comment (georgesittas) [reply sqlglot/parser.py]: ah may need to actually `self._advance()` before appending `self._prev.text.upper()` @viethungle-vt1401, sorry for this (it seems like it's related to the mypy issue)

Comment (viethungle-vt1401) [reply sqlglot/parser.py]: I tried specifying `return None`for `_advance()`, but that still didn't satisfy the linter, so going with your approach!

File: sqlglot/parser.py (hunk 18 above)
Comment (georgesittas) [reply sqlglot/parser.py]: I wouldn't bother covering these edge cases, doesn't seem worth it. SQLGlot in general doesn't function as a SQL validator so we tend to avoid introducing complexity when possible, especially for these obscure cases.

File: sqlglot/parser.py [hunk 20]
```diff
@@ -1362,9 +1362,13 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
             # exp.Properties.Location.POST_SCHEMA ("schema" here is the UDF's type signature)
             extend_props(self._parse_properties())
 
-            self._match(TokenType.ALIAS)
+            expression = None
+            if self._match(TokenType.ALIAS):
+                expression = self._parse_heredoc()
```
Comment (georgesittas) [inline sqlglot/parser.py:1367]:
```suggestion
            expression = self._match(TokenType.ALIAS) and self._parse_heredoc()
```

File: sqlglot/parser.py (hunk 18 above)
Comment (georgesittas) [reply sqlglot/parser.py]: Whatever actually, let's leave it like it is now, should be ok

**Review by georgesittas — COMMENTED** [review]
Ok one minor final comment here.

@viethungle-vt1401 let's test the `raise_error` code path, as well as the

```python
            if expression:
                extend_props(self._parse_properties())
```

one. If we can't yet generate these props correctly (according to the Slack discussion), I'd say either fix this here or just drop this change in the parser altogether.

**Comment by pkit** [comment]
Why not just?
```diff
diff --git a/sqlglot/dialects/databricks.py b/sqlglot/dialects/databricks.py
index 8e55b6a9..4258b563 100644
--- a/sqlglot/dialects/databricks.py
+++ b/sqlglot/dialects/databricks.py
@@ -81,3 +81,5 @@ class Databricks(Spark):
 
     class Tokenizer(Spark.Tokenizer):
         HEX_STRINGS = []
+        HEREDOC_STRINGS = ["$"]
+        VAR_SINGLE_TOKENS = {"$"}
```
```python
print(repr(sqlglot.parse_one("""
  CREATE FUNCTION add_one(x INT) RETURNS INT LANGUAGE PYTHON AS $FOO$def add_one(x):\n  return x+1$FOO$
""", read="databricks")))

Create(
  this=UserDefinedFunction(
    this=Identifier(this=add_one, quoted=False),
    expressions=[
      ColumnDef(
        this=Identifier(this=x, quoted=False),
        kind=DataType(this=Type.INT, nested=False, prefix=False))],
    wrapped=True),
  kind=FUNCTION,
  expression=RawString(this=def add_one(x):
        return x+1),
  exists=False,
  properties=Properties(
    expressions=[
      ReturnsProperty(
        this=DataType(this=Type.INT, nested=False, prefix=False)),
      LanguageProperty(
        this=Identifier(this=PYTHON, quoted=False))]),
  end=False)

```

**Comment by pkit** [comment]
Ah, I get it, it's because round-trip is impossible and tag is lost here: https://github.com/tobymao/sqlglot/blob/70b280f24df0f4a0d4cc8262d72ce46412f76be3/sqlglotrs/src/tokenizer.rs#L403C21-L403C24
Lol
Probably need to return the tag too...

**Comment by viethungle-vt1401** [comment]
> Ok one minor final comment here.
> 
> @viethungle-vt1401 let's test the `raise_error` code path, as well as the
> 
> ```python
>             if expression:
>                 extend_props(self._parse_properties())
> ```
> 
> one. If we can't yet generate these props correctly (according to the Slack discussion), I'd say either fix this here or just drop this change in the parser altogether.

If that's the case, I would drop the change in the parser and leave it for future updates. What's your verdict on this?

**Comment by georgesittas** [comment]
> > Ok one minor final comment here.
> > @viethungle-vt1401 let's test the `raise_error` code path, as well as the
> > ```python
> >             if expression:
> >                 extend_props(self._parse_properties())
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > one. If we can't yet generate these props correctly (according to the Slack discussion), I'd say either fix this here or just drop this change in the parser altogether.
> 
> If that's the case, I would drop the change in the parser and leave it for future updates. What's your verdict on this?

I'd just drop the logic in the parser for now and care about it in the future.

**Review by georgesittas — APPROVED** [review]
Looks good, nice work 👍
